### PR TITLE
Add verification link

### PR DESCRIPTION
**Pull request by nanaya**
Click the link and the open verification box will be magically verified. Or something.

### CODE REVIEW

File: app/Http/Middleware/VerifyPrivilegedUser.php [hunk 1]
```diff
@@ -24,6 +24,6 @@ class VerifyPrivilegedUser extends VerifyUser
 {
     public function requiresVerification($request)
     {
-        return $this->auth->user() === null || $this->auth->user()->isPrivileged();
+        return session()->get('requires_verification') ?? optional(auth()->user())->isPrivileged();
```
Comment (notbakaneko) [inline app/Http/Middleware/VerifyPrivilegedUser.php]: need some way of invalidating `requires_verification` on privilege change

File: app/Events/UserSessionEvent.php [hunk 2]
```diff
@@ -22,28 +22,39 @@
...
 
-class UserLogoutEvent extends NotificationEventBase
+class UserSessionEvent extends NotificationEventBase
 {
     public $action;
+    public $data;
     public $userId;
-    public $channelName;
+
+    public static function onLogout($userId, $keys)
```
[the remark this answers is not in the document]
Comment (nanaya) [reply app/Events/UserSessionEvent.php]: Can use some suggestion :dancer:

File: app/Libraries/Session/Store.php [hunk 3]
```diff
@@ -194,7 +193,7 @@ public function currentUserSessions()
             $sessionMeta[$id]['device'] = $agent->device();
             $sessionMeta[$id]['platform'] = $agent->platform();
             $sessionMeta[$id]['browser'] = $agent->browser();
-            $sessionMeta[$id]['verified'] = isset($session['verified']) && $session['verified'] === UserVerification::VERIFIED;
+            $sessionMeta[$id]['verified'] = (bool) ($session['verified'] ?? false);
```
[the remark this answers is not in the document]
Comment (nanaya) [reply app/Libraries/Session/Store.php]: Preventing failure when the key doesn't exist

File: app/Libraries/UserVerificationState.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        $session->start();
+
+        return new static([
+            'user' => User::find($params['userId']),
+            'session' => $session,
+            'legacySessionQueryWhere' => $params['legacySessionQueryWhere'],
+        ]);
+    }
+
+    public function __construct(array $params)
```
[the remark this answers is not in the document]
Comment (nanaya) [reply app/Libraries/UserVerificationState.php]: One of the variable name is long.

File: app/Http/Middleware/VerifyPrivilegedUser.php (hunk 1 above)
Comment (nanaya) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: Not sure how to cleanly handle that on notification server side.

Probably easier to just kick out the user's session.

File: app/Events/UserSessionEvent.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (notbakaneko) [reply app/Events/UserSessionEvent.php]: `newLogout`?

File: app/Http/Middleware/VerifyPrivilegedUser.php (hunk 1 above)
Comment (notbakaneko) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: if it's for the notification server, the web side shouldn't need to rely on the session value?

Comment (nanaya) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: Both should rely on same thing...?

Comment (notbakaneko) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: one checks per-request and the other is a persistent connection, I don't think it should be considered the same?

Comment (nanaya) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: More like per session...

Comment (notbakaneko) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: moving a logged in user to a privileged group currently triggers a verification on the next request, but this change makes it no longer trigger the verification

Comment (peppy) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: Not sure this is a real issue, to be honest.

Comment (notbakaneko) [reply app/Http/Middleware/VerifyPrivilegedUser.php]: Either:
- Privilege changes should force logout a user,
- the session value should be invalidated,
- or the session value should not be used

**Pull request merged**